File: android/app/src/main/kotlin/org/fgsec/halogen/features/settings/AccountsView.kt
package org.fgsec.halogen.features.settings

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Switch
import androidx.compose.material3.SwipeToDismissBox
import androidx.compose.material3.SwipeToDismissBoxValue
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberSwipeToDismissBoxState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.launch
import org.fgsec.halogen.components.halogenIcon
import org.fgsec.halogen.core.FriendlyError
import org.fgsec.halogen.core.HalogenCore
import org.fgsec.halogen.storage.Session

private enum class AccountsRoute { AddEmbeddedUser, AdminUsers }

/// Settings → Accounts: every saved session, switch/remove, add another
/// account (remote login or a new embedded user) — the web's accounts page.
@Composable
fun AccountsView(core: HalogenCore, onBack: () -> Unit) {
    var route by remember { mutableStateOf<AccountsRoute?>(null) }

    when (route) {
        AccountsRoute.AddEmbeddedUser -> AddEmbeddedUserView(core, onBack = { route = null })
        AccountsRoute.AdminUsers -> AdminUsersView(core, onBack = { route = null })
        null -> AccountsList(core, onBack, push = { route = it })
    }
}

@Composable
private fun AccountsList(core: HalogenCore, onBack: () -> Unit, push: (AccountsRoute) -> Unit) {
    var registry by remember { mutableStateOf(core.sessionStore.load()) }
    var confirmRemove by remember { mutableStateOf<Session?>(null) }
    val scope = rememberCoroutineScope()

    LaunchedEffect(Unit) { registry = core.sessionStore.load() }

    fun remove(session: Session) {
        if (session.id == registry.activeId) {
            core.signOut()
        } else {
            core.sessionStore.remove(session.id)
            registry = core.sessionStore.load()
        }
    }

    SettingsScaffold(title = "Accounts", onBack = onBack) { padding ->
        Column(
            Modifier.fillMaxSize().padding(padding).verticalScroll(rememberScrollState()),
        ) {
            SettingsGroup(header = "Accounts on this device") {
                for (session in registry.sessions) {
                    key(session.id) {
                        // Destructive parity: every other destructive action
                        // in the app confirms first.
                        val dismissState = rememberSwipeToDismissBoxState(
                            confirmValueChange = { value ->
                                if (value == SwipeToDismissBoxValue.EndToStart) {
                                    confirmRemove = session
                                }
                                // Never dismiss — the confirm dialog owns the removal.
                                value == SwipeToDismissBoxValue.Settled
                            },
                        )
                        SwipeToDismissBox(
                            state = dismissState,
                            enableDismissFromStartToEnd = false,
                            backgroundContent = {
                                Box(
                                    Modifier.fillMaxSize()
                                        .background(MaterialTheme.colorScheme.error),
                                    contentAlignment = Alignment.CenterEnd,
                                ) {
                                    Icon(
                                        halogenIcon("trash"),
                                        contentDescription = "Remove",
                                        tint = MaterialTheme.colorScheme.onError,
                                        modifier = Modifier.padding(end = 20.dp),
                                    )
                                }
                            },
                        ) {
                            Surface(color = MaterialTheme.colorScheme.surfaceContainerLow) {
                                SettingsRow(onClick = {
                                    if (session.id != registry.activeId) {
                                        // App scope: the switch remounts the
                                        // tree, which would cancel a
                                        // composition-scoped launch mid-way.
                                        core.scope.launch { core.switchAccount(session) }
                                    }
                                }) {
                                    Icon(
                                        halogenIcon(
                                            if (session.kind == Session.Kind.EMBEDDED) "iphone"
                                            else "server.rack"),
                                        contentDescription = null,
                                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                                    )
                                    Column(
                                        Modifier.weight(1f),
                                        verticalArrangement = Arrangement.spacedBy(1.dp),
                                    ) {
                                        Text(session.username)
                                        Text(
                                            if (session.kind == Session.Kind.EMBEDDED) "This device"
                                            else session.serverUrl ?: "",
                                            style = MaterialTheme.typography.bodySmall,
                                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                                            maxLines = 1,
                                            overflow = TextOverflow.MiddleEllipsis,
                                        )
                                    }
                                    if (session.id == registry.activeId) {
                                        Icon(
                                            halogenIcon("checkmark"),
                                            contentDescription = "Active",
                                            tint = MaterialTheme.colorScheme.primary,
                                            modifier = Modifier.size(20.dp),
                                        )
                                    }
                                }
                            }
                        }
                    }
                }
            }

            SettingsGroup(
                footer = "Switching accounts swaps the whole library view; each account keeps its own cache and settings. Removing an account signs it out on this device — nothing is deleted on the server.",
            ) {
                SettingsRow(onClick = { core.beginAddAccount() }) {
                    Icon(
                        halogenIcon("plus"),
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.primary,
                    )
                    Text("Add account (server login)", Modifier.weight(1f))
                }
                if (core.isEmbeddedAccount) {
                    SettingsRow(onClick = { push(AccountsRoute.AddEmbeddedUser) }) {
                        Icon(
                            halogenIcon("person.badge.plus"),
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.primary,
                        )
                        Text("New user on this device", Modifier.weight(1f))
                        RowChevron()
                    }
                }
                // Admin-only: manage every server user (list / edit / delete).
                if (core.isAdmin) {
                    SettingsRow(onClick = { push(AccountsRoute.AdminUsers) }) {
                        Icon(
                            halogenIcon("person.2.badge.gearshape"),
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.primary,
                        )
                        Text("Manage server users", Modifier.weight(1f))
                        RowChevron()
                    }
                }
            }
        }
    }

    confirmRemove?.let { session ->
        AlertDialog(
            onDismissRequest = { confirmRemove = null },
            title = { Text("Remove this account?") },
            text = { Text("Signs the account out on this device — nothing is deleted on the server.") },
            confirmButton = {
                TextButton(onClick = {
                    confirmRemove = null
                    remove(session)
                }) { Text("Remove ${session.username}", color = MaterialTheme.colorScheme.error) }
            },
            dismissButton = {
                TextButton(onClick = { confirmRemove = null }) { Text("Cancel") }
            },
        )
    }
}

/// Create + switch to a new user on the embedded server (username only — the
/// app generates and stores the password), the web's add-embedded-user page.
@Composable
fun AddEmbeddedUserView(core: HalogenCore, onBack: () -> Unit) {
    var username by remember { mutableStateOf("") }
    var error by remember { mutableStateOf<String?>(null) }
    var saving by remember { mutableStateOf(false) }
    val scope = rememberCoroutineScope()

    suspend fun create() {
        saving = true
        try {
            core.addEmbeddedUser(username)
        } catch (e: Exception) {
            error = FriendlyError.message(e)
        } finally {
            saving = false
        }
    }

    SettingsScaffold(title = "New device user", onBack = onBack) { padding ->
        Column(
            Modifier.fillMaxSize().padding(padding).verticalScroll(rememberScrollState()),
        ) {
            SettingsGroup(
                footer = "Users on the embedded server are always administrators. No password needed — the app manages sign-in.",
            ) {
                SettingsTextField(
                    value = username,
                    onValueChange = { username = it },
                    label = "Username",
                    capitalization = KeyboardCapitalization.None,
                )
                // Visible but locked, like the web: every embedded user is an
                // admin of the on-device server (no lesser role to manage).
                SettingsRow {
                    Text("Administrator", Modifier.weight(1f))
                    Switch(checked = true, onCheckedChange = null, enabled = false)
                }
            }
            error?.let {
                FootnoteText(
                    it,
                    color = MaterialTheme.colorScheme.error,
                    modifier = Modifier.padding(horizontal = 32.dp),
                )
            }
            Box(Modifier.padding(16.dp)) {
                BusyButton(
                    title = "Create and switch",
                    busy = saving,
                    enabled = username.trim().isNotEmpty(),
                ) {
                    // App scope: success switches accounts and remounts the
                    // tree — a composition scope cancels the POST mid-flight.
                    core.scope.launch { create() }
                }
            }
        }
    }
}

File: ios/Halogen/Features/Settings/AccountsView.swift
import SwiftUI

/// Settings → Accounts: every saved session, switch/remove, add another
/// account (remote login or a new embedded user) — the web's accounts page.
struct AccountsView: View {
    let core: HalogenCore

    @State private var registry = SessionStore.load()
    @State private var confirmRemove: Session?

    var body: some View {
        List {
            Section("Accounts on this device") {
                ForEach(registry.sessions) { session in
                    Button {
                        guard session.id != registry.activeId else { return }
                        Task { await core.switchAccount(session) }
                    } label: {
                        HStack(spacing: 12) {
                            Image(
                                systemName: session.kind == .embedded
                                    ? "iphone" : "server.rack"
                            )
                            .foregroundStyle(.secondary)
                            VStack(alignment: .leading, spacing: 1) {
                                Text(session.username).foregroundStyle(.primary)
                                Text(
                                    session.kind == .embedded
                                        ? "This device" : (session.serverUrl ?? "")
                                )
                                .font(.caption)
                                .foregroundStyle(.secondary)
                                .lineLimit(1)
                                .truncationMode(.middle)
                            }
                            Spacer()
                            if session.id == registry.activeId {
                                Image(systemName: "checkmark")
                                    .foregroundStyle(Color.accentColor)
                            }
                        }
                    }
                    .swipeActions(edge: .trailing) {
                        // Destructive parity: every other destructive action
                        // in the app confirms first.
                        Button(role: .destructive) {
                            confirmRemove = session
                        } label: {
                            Label("Remove", systemImage: "trash")
                        }
                    }
                }
            }

            Section {
                Button {
                    core.beginAddAccount()
                } label: {
                    Label("Add account (server login)", systemImage: "plus")
                }
                if case .embedded = core.account?.kind {
                    NavigationLink {
                        AddEmbeddedUserView(core: core)
                    } label: {
                        Label("New user on this device", systemImage: "person.badge.plus")
                    }
                }
                // Admin-only: manage every server user (list / edit /
                // delete) — the web's Accounts "Manage" button.
                if core.isAdmin {
                    NavigationLink {
                        AdminUsersView(core: core)
                    } label: {
                        Label("Manage server users", systemImage: "person.2.badge.gearshape")
                    }
                }
            } footer: {
                Text("Switching accounts swaps the whole library view; each account keeps its own cache and settings. Removing an account signs it out on this device — nothing is deleted on the server.")
            }
        }
        .navigationTitle("Accounts")
        .navigationBarTitleDisplayMode(.inline)
        .onAppear { registry = SessionStore.load() }
        .confirmationDialog(
            "Remove this account?",
            isPresented: Binding(
                get: { confirmRemove != nil },
                set: { if !$0 { confirmRemove = nil } }
            )
        ) {
            if let session = confirmRemove {
                Button("Remove \(session.username)", role: .destructive) {
                    remove(session)
                }
            }
        } message: {
            Text("Signs the account out on this device — nothing is deleted on the server.")
        }
    }

    private func remove(_ session: Session) {
        if session.id == registry.activeId {
            core.signOut()
        } else {
            SessionStore.remove(session.id)
            registry = SessionStore.load()
        }
    }
}

/// Create + switch to a new user on the embedded server (username only — the
/// app generates and stores the password), the web's add-embedded-user page.
struct AddEmbeddedUserView: View {
    let core: HalogenCore

    @State private var username = ""
    @State private var error: String?
    @State private var saving = false

    var body: some View {
        Form {
            Section {
                TextField("Username", text: $username)
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)
                // Visible but locked, like the web: every embedded user is an
                // admin of the on-device server (no lesser role to manage).
                Toggle("Administrator", isOn: .constant(true))
                    .disabled(true)
            } footer: {
                Text("Users on the embedded server are always administrators. No password needed — the app manages sign-in.")
            }
            if let error {
                Text(error).font(.footnote).foregroundStyle(.red)
            }
            Button {
                Task { await create() }
            } label: {
                if saving {
                    ProgressView().frame(maxWidth: .infinity)
                } else {
                    Text("Create and switch").frame(maxWidth: .infinity)
                }
            }
            .disabled(saving || username.trimmingCharacters(in: .whitespaces).isEmpty)
        }
        .navigationTitle("New device user")
        .navigationBarTitleDisplayMode(.inline)
    }

    private func create() async {
        saving = true
        defer { saving = false }
        do {
            try await core.addEmbeddedUser(username: username)
        } catch {
            self.error = FriendlyError.message(error)
        }
    }
}
